### ISSUE TITLE
function signature for GrB_IndexUnaryOp

### ISSUE THREAD

**Issue by DrTimothyAldenDavis** [GraphBLAS/graphblas-api-c#33]
The new draft spec has a GrB_IndexUnaryOp, which will be used in GrB_apply and GrB_select.  The underlying function has the following proposed signature:

`void func (void *result, const void *aij, GrB_Index *indices, int n, const void *thunk) ;`

The indices array has size n = 1 for vectors or n = 2 for matrices.  For a matrix, it holds [i j], the row and column index of the entry A(i,j).

I see many difficult problems with this.

It will be very difficult to do on the GPU, which will severely affect the performance of user-defined IndexUnaryOps on the GPU.

On the GPU, creating teeny arrays (like the GrB_Index indices [2] array) cannot be done per thread.  A bulk array would have been created in shared memory, and the [i j] indices loaded there.  The GPU does much better with individual scalars say i and j, which can be placed in per-thread registers.  Consider this example for a parallel apply to a CSC matrix:

in an outer loop:

`int64_t j = ... ;  //the current column index.  The GPU can put this in register.`

in an inner loop, where a single thread loads its index i:
```
int64_t i = Ai [p + threadIdx.x] ;  // the GPU can put this in register
func ( ..., i, j, ... ) ;   // apply the function
```

If instead, these indices i and j must be placed into an indices array of size 2, the only way to do this is to create a shared array of size # threads times 2, and then threadId access its component of size 2.  This shared array cannot be in registers but in slower shared memory.

It just seems like there's unnecessary data movement to move the i and j indices from where the come from in the matrix data structure, to pack them into this teeny indices array of size 2, for each call to the function.

I would prefer the following function signature, where j = 0 is passed if the function is applied to vectors:

`void func (void *result, const void *aij, int64_t i, int64_t j, const void *thunk) ;`

This also avoids another problem.  Currently, I save a vast amount of code by treating a GrB_Vector as an n-by-1 GrB_Matrix, internally.  The user doesn't know or see this, but it allows me to write just methods that work a GrB_Matrix.

With the new GrB_IndexUnaryOp, I now would have to keep track of the difference between a GrB_Vector and an n-by-1 GrB_Matrix, which is doable but painful.  I would instead just pass n = 2 all the time, and never pass n=1.  For GrB_Vectors I would just pass in indices as [i  0].  This would circumvent the problem but not in a portable way.

I also see a useless branch that is likely in user-defined operators.  Since the indices array has size 1 or 2, the user will need to test it and write code like:

```
void func ( .... indices, n, ...)
int64_t j = (n == 1) ? 0 : indices [1] ;
result = some function of aij, i, j, the thunk, etc.
```

That branch in the ternary is slow on the GPU, which has simple cores with no pipelined instructions.

What happens if a method defined for just matrices, like TRIL, is applied to a GrB_Vector?  An error is returned?  If so, then shouldn't the user be able to define a GrB_IndexUnaryOp that also returns an error if it is applied to a GrB_Vector?  That adds an extra layer of awkwardness than can be entirely avoided by ditching the parameter n, and passing in i and j always, even for vectors.  Just pass in j as zero, if working on a GrB_Vector.

Finally, GrB_Index is a bad idea.  It makes user-defined operators that act like DIAGINDEX very difficult.  Say I want to create a user-defined operator "TRIDIAG" which returns true for entries on the -1, 0, and +1 diagonals, and false otherwise.  Then I could write that as:

```
int64_t d = (i-j) ;
result = (d >= -1 && d <= 1) ;
```

That computation cannot be done in uint64_t.

In summary, I see many problem with the parameters 

`..., GrB_Index *indices, GrB_Index n, ...`

and suggest that they instead be

`..., int64_t i, int64_t j, ...`

**Comment by DrTimothyAldenDavis** [comment]
Also, the UINT32 and UINT64 versions of the built-in methods should not be added.  Subtracting indices of UINT32 or UINT64 can lead to bizarre results, like 0-1 being returned as UINT64_MAX, which cannot be portablly typecasted to negative one.

DIAGINDEX_* is backwards, the result should be (j - (i+s)), so that positive diagonals are above the main diagonal, like they are in matrix algebra.

**Comment by DrTimothyAldenDavis** [comment]
Consider TRIL.  If you want to get entries far below the main diagonal, you can pass in thunk as negative.  But that cannot be done for GrB_TRIL_UINT64, where all inputs (including the thunk) are unsigned integers.

All integer computations on indices must be done in signed arithematic.  This does require the largest index to be no bigger than INT64_MAX, of course.  I have a GxB_INDEX_MAX that is set to 2^60, which ensures this (GrB_INDEX_MAX should be added to the spec; see the other issue about that).

**Comment by mcmillan03** [comment]
Tim Mattson is investigating further.  Scott is going to try to figure out what the table in Chapter 3 would look like.

**Comment by DrTimothyAldenDavis** [comment]
Why are the indices passed in as uint64_t?  They should be int64_t since index computations such as i-j must be done with signed integers, not unsigned.

**Comment by mcmillan03** [comment]
1. I understand the simplicity of passing 0...and the tension between math and software engineering (I will strive to "make it hard for the user to do the wrong thing wherever possible"). This is still an open discussion.
2. We already understand the issues with 'n' (stated in the previous comment).
3. The union is only one type (not two).  Since I can do what is proposed in GBTL without doubling the implementations, I can only assume this is a byproduct of your implementation that does not distinguish between matrices and vectors.  Or you could always set j to zero and have implementation defined behaviour that is equivalent to this proposal regardless. (less satisfying, I know).
4. The types of the parameters and what type the operations are performed in (two different things) is the topic of a different issue: #45

**Comment by DrTimothyAldenDavis** [comment]
The part that is confusing to me is that currently all operators work the same on both matrices, vectors, and even scalars (the latter with the accum operator in GrB-reduce, for example).  This is the first place where an operator will simply break if an attempt is made to use it on a vector.

What happens with GrB_Vector_select when using a GrB_IndexUnaryOp only defined for matrices, like GrB_TRIL, GrB_COLINDEX, etc?  The spec states

>  Operators requiring access to two indices only work with GrB_Matrix and will result in undefined behavior if used with GrB_Vector.

That's not very satisfying.  "undefined behavior" can be a segfault.  It would be so much simpler just to state that vector entry v(i) has an index i, and that j is simply zero.

Undefined behavior makes my testing suite more difficult.   If this is left in, I will just never use the GrB_Vector variant of these ops, and always pass in j = 0.  My "undefined behavior" will be nicely and simply defined.

**Comment by DrTimothyAldenDavis** [comment]
I don't understand the issue regarding the math vs software engineering distinction.  We already live under the domain of software engineering when it comes to matrices and vectors in GraphBLAS.  Consider the way a vector is created:

```
GrB_Index n  = (some expression) ;
GrB_Vector_new (&v, GrB_FP64, n) ;
```

In math, n is the vector length, and for a true vector, n is a member of the set natural numbers, which is infinite.  In GraphBLAS, n comes from a  finite set, 1 to 2^(64)-1.  In addition, v is not a real vector, but we pretend it is.   Its data type comes from a finite set of rational numbers (double).  These restrictions are far more limiting than defining the index of a vector in a function signature.  The definition is well-defined mathematically:

(1) for matrices, i is the row index and j is the column index
(2) for vectors, i is the index, and j is not an index, but is zero to denote that the input is a vector.

Why make the function signature hard to evaluate, slow on the GPU, and harder to code on the CPU, when a simpler solution exists which is also faster?

**Comment by DrTimothyAldenDavis** [comment]
What is the decision on the final signature?  Is this it?

`void func (void *result, const void *aij, GrB_Index i, GrB_Index j, const void *s) ;`

**Comment by mcmillan03** [comment]
That is the signature.  Jose is working out the issues related to any casting that would be required (or not) inside the function

**Comment by DrTimothyAldenDavis** [comment]
Great!  I'm working on updating LAGraph to require the new v2.0 C API.  This will go a long way to combining the vanilla and GxB* SuiteSparse variants into a single vanilla variant, with GrB_select being used throughout.

**Comment by DrTimothyAldenDavis** [comment]
Lines 1264-1267 still show the old signature, but this issue is closed:

```
void func(void *out,
          const void *in1, GrB_Index* indices, GrB_Index n,
          const void *in2);
```

Doesn't this need updating?  Also see the signature in the GrB_IndexUnaryOp_new method.

**Comment by mcmillan03** [comment]
On it...
